### ISSUE TITLE
Add static method to create an IEnumerable with one element

### ISSUE THREAD

**Issue by benjamin-hodgson** [dotnet/corefx#3093]
I find that I need the following static method quite often:

```csharp
public static class MoreEnumerableMethods
{
    // ...
    public static IEnumerable<T> Singleton<T>(T value)
    {
        return Enumerable.Repeat(value, 1);
    }
}
```

This seems like a really good fit for `Enumerable` itself, where it would live alongside `Repeat`, `Empty`, and `Range`. I want to be able to say `Enumerable.Singleton(x)`, not `MoreEnumerableMethods.Singleton(x)`.

Even Eric Lippert has [complained](http://ericlippert.com/2013/03/18/monads-part-eight/) about the lack of this method:
> I wish there was a static method on Enumerable specifically for making a one-element sequence.

I have the code (and tests) ready on [a branch in my fork](https://github.com/benjamin-hodgson/corefx/tree/singleton). I can open a pull request right away, subject to any required API reviews.

**Comment by JonHanna** [comment]
I call the version I often use, `Emit`, because I think of calling it as an extension as "emitting" it into an enumerable. *Emit* has some other meanings of course, but perhaps not quite as strongly hit as *Singleton*.

I tend to have it as an extension thus:

    public static IEnumerable<T> Emit<T>(T element)
    {
      return Repeat(element, 1);
    }

**Comment by JonHanna** [comment]
That said, having it as an extension already solves the irritation as to how it is called: Change yours to be an extension and then it's `x.Singleton()`.

**Comment by benjamin-hodgson** [comment]
@JonHanna I actually prefer a static method to an extension in this case. For one, it's more cohesive with similar methods on `Enumerable` such as `Range` and `Repeat`. For another, putting a single element in a list doesn't feel to me like something you _do to the element_. It's a class-level operation of `Enumerable`, not an operation with which we want to extend an arbitrary object - the active party is the `Enumerable` class, not the object. I think it would look a bit strange with the extension-method syntax.

**Comment by JonHanna** [comment]
You could say the same for every `AsXXX` method.

**Comment by benjamin-hodgson** [comment]
Well, yes. I actually would have preferred `List.FromEnumerable` to `Enumerable.ToList`. But it's too late (and not important enough) to complain about that, those methods have already been written.

It seems pretty clear to me that for consistency with closely related methods like `Repeat`, it has to be a static method on `Enumerable`.

**Comment by JonHanna** [comment]
It's not clear to me why it should be consistent with `Repeat`. It's acting on a single element in a way that `Repeat` is not, where the element and the count don't have a clear primacy.

**Comment by benjamin-hodgson** [comment]
I can see us debating this until the cows come home. I think we need some further input on this question from a repo owner.

**Comment by benjamin-bader** [comment]
From my perspective (the peanut gallery, to be sure), `Emit` (or `Singleton` or whatever) makes more sense as a static `Enumerable` member, for the reason that its job is constructing `IEnumerable`s.  There's abundant precedent here - `Enumerable` is full of static factory methods e.g. `Range`; it is the natural peg on which to hang such methods.

My personal preference for the actual method would be something like
```csharp
Enumerable.Of<T>(params T[] items)
{
    return items;
}
```

...though I'm sure it could be improved upon, for example with respect to heap allocations.

**Comment by stephentoub** [comment]
FWIW, I personally don't find a need for this method.  Any time I've wanted this, I've just done ```new[]{item}```.  Short of that, there's the ```Enumerable.Repeat(item,1)``` already mentioned.  To me, while the need for such functionality certainly arises from time to time, either of those solutions seems sufficient for the frequency with which I need it.

**Comment by JohanLarsson** [comment]
I agree that `Enumerable.Single(item)` would be nice while `Enumerable.Repeat(item,1)` does the same thing single would be a bit clearer. An alternative name could be`Enumerable.One(item)`.

**Comment by karelz** [comment]
What about adding `Enumerable.Create<T>(T item)`, with overloads like `.Create<T>(T item1, T item2)` up to 5-8 overloads. That would be even more useful.
Would everyone be fine with the API name `Create<T>`?

cc: @VSadov @OmarTawfik

**Comment by jnm2** [comment]
@stephentoub Interestingly I've toyed with the idea of an allocation-free `EnumerableEx.Single<T>(T value)`. Not sure if that would be interesting. I do run into `new[] { x }` not infrequently.

**Comment by stephentoub** [comment]
> I've toyed with the idea of an allocation-free EnumerableEx.Single<T>(T value)

If EnumerableEx.Single would return an ```IEnumerable<T>```, how would it be allocation-free?  Are you thinking of caching instances for specific T values?

**Comment by svick** [comment]
@karelz I think I would prefer a name that works well with `using static System.Linq.Enumerable;`.

For comparison, looking at whether the existing non-extension methods on `Enumerable` work in that situation, I'd say that `Range()` is good, `Repeat()` is acceptable, only `Empty()` is bad.

**Comment by jnm2** [comment]
@stephentoub It would return a struct which implements IEnumerable but has GetEnumerator return another struct, etc.

**Comment by stephentoub** [comment]
> @stephentoub It would return a struct which implements IEnumerable<> but has GetEnumerator return another struct, etc.

Oh, you're suggesting its signature actually include the struct rather than IEnumerable<T> as the return type?  I'm not clear on what scenarios this really helps.  Presumably the reason you want an enumerable for a single object is because you're going to be using it as part of other code that works with enumerables, at which point you're going to have to cast this to IEnumerable<T> and box it.  If that's not the case, then why wouldn't you just pass around the underlying single instance rather than the struct that wraps it?

**Comment by karelz** [comment]
> @svick For comparison, looking at whether the existing non-extension methods on `Enumerable` work in that situation, I'd say that `Range()` is good, `Repeat()` is acceptable, only `Empty()` is bad.

I have no idea why you rated them this way. Can you please shed some light here? Why is `Empty` worse than `Range`? Maybe post a code snippet demonstrating your point may help me understand ...

**Comment by svick** [comment]
@karelz 

If I saw code like `var range = Range(1, 10);`, I would expect `range` to be sequence of consecutive numbers.

If I saw `var repeated = Repeat(-1, 10);`, I would expect `repeated` to be a sequence of -1 repeated 10 times.

If I saw `var empty = Empty<int>();`, I wouldn't know what to expect.

Likewise, if I saw `var created = Create(1);`, I wouldn't know what to expect.

My point is that the names `Range` and `Repeat` on their own imply sequences (at least to me), `Empty` and `Create` don't.

**Comment by karelz** [comment]
@svick I see your point now, thanks for explanation. Is there any special reason why `Enumerable` should play well in `using` statements on the class name?
I personally never used `using` on a class (class name as context is great in the code IMO). Naively, I don't believe our methods in general are well named at all to support this pattern. I don't think we should even go after that goal - that would just often make method naming repeat the context of its class IMO.

I would personally focus on `Enumerable.Create(1,2)` ... that should give you idea what it does.
Just my 2 cents (I'll be happy to adapt my view point if enough people think `using Enumerable` is a common pattern).

**Comment by svick** [comment]
@karelz

> Is there any special reason why `Enumerable` should play well in using statements on the class name?

I think that it is something that's nice to have, especially for types that already work at least somewhat with `using static` and that are already used with it (see [this GitHub search](https://github.com/search?p=2&q=%22using+static+System.Linq.Enumerable%22&type=Code) for some examples).

But I agree that repeating context is not a good idea. And I don't have a good alternative name. (The best I can think of is to steal [`Singleton` from F#](https://msdn.microsoft.com/en-us/visualfsharpdocs/conceptual/seq.singleton['t]-function-[fsharp]), but that wouldn't be great choice for several reasons.)

**Comment by karelz** [comment]
More used in `using` than I expected, thanks for the data @svick!

So far the best name is `Create`, if someone comes up with better name, we can consider it (`Singleton` is not best choice).

**Comment by jnm2** [comment]
@stephentoub 
> If that's not the case, then why wouldn't you just pass around the underlying single instance rather than the struct that wraps it?

Yeah, looks like there's nothing here. My brain hadn't reached the conclusion. So next question is, what's the least expensive allocation? Probably `T[]` is as cheap as any.

**Comment by jnm2** [comment]
I like the name Single a lot, especially when a single item is going in a slot where multiple items are expected.

**Comment by JonHanna** [comment]
`Single` is already a name of a method on `Enumerable`. Even if it didn't clash due to overloads it still clashes mnemonically.
I think I like `Create` the best myself.

**Comment by stephentoub** [comment]
> Yeah, looks like there's nothing here. My brain hadn't reached the conclusion. So next question is, what's the least expensive allocation? Probably T[] is as cheap as any.

@jnm2, the smallest possible allocation on 64-bit is 24 bytes: 16 bytes for the object header, and then a minimum of 8 bytes for the state of the object.  So a T[1] where T is <= 8 bytes would be 24 bytes.  However, if you're asking as part of thinking through an implementation of the method being discussed here, almost certainly you'd want the object to be a custom one that implements ```IEnumerable<T>``` and ```IEnumerator<T>```, so that same object can store the value, be returned from the method, and also be reused as the enumerator from GetEnumerator.  This is how many of the custom enumerators in LINQ work today.

> So far the best name is Create, if someone comes up with better name, we can consider it (Singleton is not best choice).

FWIW, I'm not a fan of Create.  The majority of the methods on Enumerable create something, whether it's a creating an enumerable that repeats a certain value a number of times, or creates an enumerable that produces a range, or creates an enumerable that maps from one input to another, etc.  Thus, using Create only for the one that handles a single value seems inconsistent to me.

How about one of these?
```C#
public static IEnumerable<T> Wrap<T>(T item);
public static IEnumerable<T> Wrap<T>(T item1, T item2, ...);

public static IEnumerable<T> Yield<T>(T item);
public static IEnumerable<T> Yield<T>(T item1, T item2, ...);

public static IEnumerable<T> Items<T>(T item);
public static IEnumerable<T> Items<T>(T item1, T item2, ...);

public static IEnumerable<T> Elements<T>(T item);
public static IEnumerable<T> Elements<T>(T item1, T item2, ...);

public static IEnumerable<T> Echo<T>(T item);
public static IEnumerable<T> Echo<T>(T item1, T item2, ...);

public static IEnumerable<T> Lift<T>(T item);
public static IEnumerable<T> Lift<T>(T item1, T item2, ...);
```

**Comment by karelz** [comment]
Good point from @stephentoub that majority in `Enumerable` creates something. I like `Elements` and `Items`.

**Comment by benjamin-bader** [comment]
Perhaps my years on the JVM are showing, but I still like good old `Of`: concise and descriptive:

```
public static IEnumerable<T> Of<T>(T item);

// Seems pretty clear to me!
var anEnum = Enumerable.Of("foo");
```

**Comment by VSadov** [comment]
I think Create is good here.  It is true that many other methods create something, but the typical use will be:    "Enumerable.Create(...)"   which is clearly creating an enumerable.

**Comment by VSadov** [comment]
It seems like a good addition.
Before this is implemented we  need to agree on
- The name of the factory
- How many overloads  ( 1-8, N)
- Do we need a "params" overload
- Pass the above through API review.

**Comment by JonHanna** [comment]
A params overload wouldn't add much, since params itself is just sugar on an array, but it would be intuitive that there was no arbitrary limit.
Would just returning that array be considered leaking interfaces?

**Comment by VSadov** [comment]
@JonHanna it would need at least to wrap the array into something so it could not be cast back to an array  and changed.

I am not very convinced that params version is needed.

**Comment by VSadov** [comment]
As a minimal API change that would address the issue we only need something like

```cs
        static IEnumerable<T> CreateFrom<T>(T element)
        {
            yield return element;
        }
```

**Comment by JonHanna** [comment]
I've we're going to insist that the params must be wrapped, then I'm against the params form.

**Comment by jamesqo** [comment]
I think we should hold on this. There is an issue in Roslyn, https://github.com/dotnet/roslyn/issues/36, that will allow for `params IEnumerable<T>` support. When that is implemented we could just have a single method:

```cs
public static IEnumerable<T> WhateverName<T>(params IEnumerable<T> items)
{
    return items;
}
```

And let the compiler do all the optimizations (e.g. caching into a static if all arguments are constants). We should definitely not add a `params T[]` which will, as @VSadov mentions, force us to make another allocation.

**Comment by karelz** [comment]
The roslyn issue is marked as "2 - Ready" for a year (https://github.com/dotnet/roslyn/issues/36#event-469859827). It does not have milestone. I don't think it is planned anytime soon.

**Comment by jamesqo** [comment]
@karelz Yes, but the language designers seemed to be generally in favor of it. If we add a `params T[]` overload and then this gets implemented in the future, adding that will have been a grave mistake.

**Comment by karelz** [comment]
The comments above already suggest `params` is not needed.

**Comment by jamesqo** [comment]
@karelz, where? I only took a quick look with Ctrl+F, but AFAICS only @JonHanna and @VSadov oppose `params`, and that's because it would have to be wrapped in an iterator with `params T[]`. Having `params IEnumerable<T>` solves that problem.

**Comment by karelz** [comment]
Yes, both @JonHanna and @VSadov oppose adding `params T[]` overload, which is the point I tried to make.
I think we can go ahead with the other overloads, unless the Roslyn feature `params IEnumerable<T>` is scheduled for next Roslyn version ...
Or did I misunderstand anything?

**Comment by jamesqo** [comment]
@karelz Sorry for the delayed response. I think it would be fine to have a non-params overload. On further thought it would allow us to make optimizations the compiler can't, like returning a cached enumerable based on the runtime value of the input.

In regards to naming, I vote for `Create` or `Of`. (I like `Of` better, but I admit that `Create` is more consistent with the naming conventions elsewhere.)

**Comment by jamesqo** [comment]
@karelz, is there anything else blocking this API? Is it ready for review?

**Comment by karelz** [comment]
We need final API proposal + agreement on the naming. Then we can submit it for approval.

**Comment by tuespetre** [comment]
FWIW I feel that @stephentoub [struck naming gold with `Yield`.](https://github.com/dotnet/corefx/issues/3093#issuecomment-260949302)

```
using static System.Linq.Enumerable;

sequence = sequence.Concat(Yield(item));
```

There are two reasons I feel that `Yield` is the perfect name:

1. The obvious familiarity of the `yield` language construct 
2. The existing use of the name `Yield` by the `Task` class, which is used to 'wrap' a value into a `Task<TResult>`, similar to how `Enumerable.Yield<TResult>` would be used to 'wrap' a value into an `IEnumerable<TResult>`

Furthermore, `Empty` and `Repeat` use the name `TResult` for their generic type parameter, so `Yield` should also use the name `TResult` for its generic type parameter.

```
public static IEnumerable<TResult> Yield<TResult>(TResult item);
```

🍰 🍰 🍰

**Comment by jamesqo** [comment]
@tuespetre `Yield` seems like a good, unambiguous name. I've changed my mind on having `Create` above since that would not work well with `using static`.

@JonHanna @svick @jnm2 and others, does this sound like an acceptable name to you? Please be aware that this API will never get through unless we reach a consensus.

**Comment by JonHanna** [comment]
I wouldn't object to `Yield`.

**Comment by jnm2** [comment]
Not my first choice but I wouldn't object either.

**Comment by Ricciolo** [comment]
My 2 cents: I like `Yield` 👍

**Comment by jamesqo** [comment]
@karelz Can we mark as ready-to-review for `Yield`? 6 people are in favor, with none opposed so far.

**Comment by karelz** [comment]
I let area experts to decide - @VSadov @OmarTawfik what do you think?

**Comment by jamesqo** [comment]
As a reminder, the current API proposal is

```cs
public static IEnumerable<T> Yield<T>(T element);
public static IEnumerable<T> Yield<T>(params IEnumerable<T> elements);
```

with the second overload slated to be released when `params IEnumerable<T>`is implemented in Roslyn.

**Comment by weitzhandler** [comment]
I see this far less important than #23583, but I'm biased.

**Comment by weitzhandler** [comment]
`Yield` sounds like a nice name to me, if I'm allowed to comment on that.

Anyway, I bump into the need to generate a series of a multiple (let alone a single) reference to a single object far less often than to generate an `IEnumerable<T>` of non-repeated objects.

#23583 is more important to me, but I'm biased. `Yield` seems like the perfect solution if I got the picture well.

I'd want a shortcut to `Enumerable.Range(10).Select(i => new Customer { Number = i })`. I bump into this really often. Please consider this.

**Comment by karelz** [comment]
@weitzhandler your suggestion #23583 should not be mingled into this issue. It is different API request with different goals.
I would suggest to reopen your original issue #23583 if you want to have a discussion about it.

**Comment by gtbuchanan** [comment]
`EnumerableEx.Return` already exists in the now .NET Foundation authored package [`System.Interactive`](https://github.com/dotnet/reactive/blob/master/Ix.NET/Source/System.Interactive/Return.cs). I think it makes sense to stick with the same naming. It aligns with `System.Reactive`'s `Observable.Return`. Related to #14119
